crypto/evp/evp_rand.c

```diff
@@ -645,12 +645,13 @@ static int evp_rand_nonce_locked(EVP_RAND_CTX *ctx, unsigned char *out,
                                  size_t outlen)
 {
     unsigned int str = evp_rand_strength_locked(ctx);
+    size_t len = 0;
 
-    if (ctx->meth->nonce == NULL)
-        return 0;
-    if (ctx->meth->nonce(ctx->algctx, out, str, outlen, outlen))
-        return 1;
-    return evp_rand_generate_locked(ctx, out, outlen, str, 0, NULL, 0);
+    if (ctx->meth->nonce != NULL)
+        len = ctx->meth->nonce(ctx->algctx, out, str, outlen, outlen);
+    else if (evp_rand_generate_locked(ctx, out, outlen, str, 0, NULL, 0) > 0)
+        len = outlen;
+    return len > INT_MAX ? INT_MAX : (int)len;
```

Comment:
I wonder if this should be documented in EVP_RAND_nonce()..
Passing a size_t and then returning a int is just horrible.